File: web/src/utils/syncMerge.ts
export function recordTimestamp(updatedAt?: number, fallback = 0): number {
  return updatedAt && updatedAt > 0 ? updatedAt : fallback;
}

export interface MergeableRecord {
  cloudId: string;
  id?: number;
  updatedAt?: number;
  deleted?: boolean;
  pendingSync?: boolean;
}

export function mergeById<T extends MergeableRecord>(
  localItems: T[],
  remoteItems: T[],
  options?: { isIncremental?: boolean; ignoreRemoteDeletes?: boolean; pendingDeleteCloudIds?: Set<string> },
): { toApplyLocal: T[]; toPushRemote: T[]; toDeleteLocal: number[] } {
  const isIncremental = options?.isIncremental ?? false;
  const ignoreRemoteDeletes = options?.ignoreRemoteDeletes ?? false;
  const pendingDeleteCloudIds = options?.pendingDeleteCloudIds ?? new Set<string>();
  const localMap = new Map(localItems.map((item) => [item.cloudId, item]));
  const remoteMap = new Map(remoteItems.map((item) => [item.cloudId, item]));
  const allCloudIds = new Set(
    isIncremental
      ? [
          ...remoteMap.keys(),
          ...localItems.filter((item) => item.pendingSync).map((item) => item.cloudId),
        ]
      : [...localMap.keys(), ...remoteMap.keys()],
  );

  const toApplyLocal: T[] = [];
  const toPushRemote: T[] = [];
  const toDeleteLocal: number[] = [];

  for (const cloudId of allCloudIds) {
    const local = localMap.get(cloudId);
    const remote = remoteMap.get(cloudId);

    if (!local && remote) {
      if (!remote.deleted && !pendingDeleteCloudIds.has(cloudId)) toApplyLocal.push(remote);
      continue;
    }

    if (local && !remote) {
      if (!isIncremental || local.pendingSync) toPushRemote.push(local);
      continue;
    }

    if (local && remote) {
      const localAt = recordTimestamp(local.updatedAt);
      const remoteAt = recordTimestamp(remote.updatedAt);

      if (remote.deleted) {
        if (!ignoreRemoteDeletes) {
          if (remoteAt >= localAt && local.id != null) toDeleteLocal.push(local.id);
          else toPushRemote.push(local);
        }
        continue;
      }

      if (remoteAt > localAt) toApplyLocal.push(remote);
      else if (localAt > remoteAt) toPushRemote.push(local);
      else toApplyLocal.push(remote);
    }
  }

  return { toApplyLocal, toPushRemote, toDeleteLocal };
}

export function mergePreferences<T extends { updatedAt: number }>(local: T, remote: T | null): T {
  if (!remote) return local;
  return remote.updatedAt > local.updatedAt ? remote : local;
}

File: app/src/main/java/com/aus/ausgegeben/sync/SyncMerge.kt
package com.aus.ausgegeben.sync

interface SyncItem {
    val id: Long
    val updatedAt: Long
    val deleted: Boolean
}

data class MergeResult<T>(
    val toApplyLocal: List<T>,
    val toPushRemote: List<T>,
    val toDeleteLocal: List<Long>,
)

fun recordTimestamp(updatedAt: Long, fallback: Long = 0L): Long =
    if (updatedAt > 0L) updatedAt else fallback

fun <T> mergeById(
    localItems: List<T>,
    remoteItems: List<T>,
    idSelector: (T) -> Long,
    updatedAtSelector: (T) -> Long,
    deletedSelector: (T) -> Boolean,
): MergeResult<T> where T : Any {
    val localMap = localItems.associateBy(idSelector)
    val remoteMap = remoteItems.associateBy(idSelector)
    val allIds = localMap.keys + remoteMap.keys

    val toApplyLocal = mutableListOf<T>()
    val toPushRemote = mutableListOf<T>()
    val toDeleteLocal = mutableListOf<Long>()

    for (id in allIds) {
        val local = localMap[id]
        val remote = remoteMap[id]

        when {
            local == null && remote != null -> {
                if (!deletedSelector(remote)) toApplyLocal += remote
            }
            local != null && remote == null -> {
                toPushRemote += local
            }
            local != null && remote != null -> {
                val localAt = updatedAtSelector(local)
                val remoteAt = updatedAtSelector(remote)
                if (deletedSelector(remote)) {
                    if (remoteAt >= localAt) toDeleteLocal += id
                    else toPushRemote += local
                } else when {
                    remoteAt > localAt -> toApplyLocal += remote
                    localAt > remoteAt -> toPushRemote += local
                    else -> toApplyLocal += remote
                }
            }
        }
    }

    return MergeResult(toApplyLocal, toPushRemote, toDeleteLocal)
}

fun mergePreferences(local: SyncedPreferences, remote: SyncedPreferences?): SyncedPreferences {
    if (remote == null) return local
    return if (remote.updatedAt > local.updatedAt) remote else local
}
